### PR TITLE
Fix PR926 local properties issues in Spark Streaming like scenarios

### PR DESCRIPTION
**Pull request by jerryshao**
[PR926](https://github.com/mesos/spark/pull/926) changes local properties from `DynamicVariable` to `ThreadLocal`,   which cannot pass properties set in parent thread to child thread. Scenarios like Spark Streaming which separate `runJob()` function in child thread will always get `null` properties and pass to DAGScheduler. So thread local properties should be inheritable, also should be isolated in different threads.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/ThreadingSuite.scala [hunk 1]
```diff
@@ -149,4 +149,40 @@ class ThreadingSuite extends FunSuite with LocalSparkContext {
...
+          assert(sc.getLocalProperty("test") === "parent")
+          sc.setLocalProperty("test", i.toString)
+          assert(sc.getLocalProperty("test") === i.toString)
+        }
+      }
+    }
+
+    threads.foreach(_.start())
+    assert(sc.getLocalProperty("test") === "parent")
+    assert(sc.getLocalProperty("Foo") === null)
```
Comment (rxin) [inline core/src/test/scala/org/apache/spark/ThreadingSuite.scala:193]: Don't we need a barrier here to properly test this?

File: core/src/test/scala/org/apache/spark/ThreadingSuite.scala [hunk 2]
```diff
@@ -149,4 +149,40 @@ class ThreadingSuite extends FunSuite with LocalSparkContext {
...
+    threads.foreach(_.start())
+
+    assert(sc.getLocalProperty("test") === null)
+  }
+
+  test("set and get local properties in parent-children thread") {
+    sc = new SparkContext("local", "test")
+    sc.setLocalProperty("test", "parent")
+
+    val threads = (1 to 5).map{ i =>
```
Comment (rxin) [inline core/src/test/scala/org/apache/spark/ThreadingSuite.scala]: Adding a space before {

File: core/src/test/scala/org/apache/spark/ThreadingSuite.scala [hunk 3]
```diff
@@ -149,4 +149,40 @@ class ThreadingSuite extends FunSuite with LocalSparkContext {
       fail("One or more threads didn't see runningThreads = 4")
     }
   }
+
+  test("set local properties in different thread") {
+    sc = new SparkContext("local", "test")
+
+    val threads = (1 to 5).map{ i =>
```
Comment (rxin) [inline core/src/test/scala/org/apache/spark/ThreadingSuite.scala]: space before {

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 4]
```diff
@@ -273,6 +275,8 @@ class SparkContext(
     }
   }
 
+  def getLocalProperty(key: String): String = Option(localProperties.get).map(_.getProperty(key)).getOrElse(null)
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Make sure the line does not exceed 100 chars.

**Comment by kayousterhout** [comment]
Thanks for fixing this -- sorry for not realizing this use case!  The fix looks good.

**Comment by jerryshao** [comment]
Hi @rxin ,thanks for your reviewing. I think it's more reasonable to add a barrier to unit test, I've already update the patch.

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
We should also merge this back into 0.8 branch once 0.8 is released.

**Comment by rxin** [comment]
I've merged this into asf spark branch-0.8.

**Pull request merged**